src/linker/Linker.Steps/MethodBodyInstructionsProvider.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Mono.Collections.Generic;
+
+namespace Mono.Linker.Steps
+{
+	public class MethodBodyInstructionsProvider
```

Comment:
Looking for naming ideas - so consider this a temporary name.

Reply:
I intentionally create a "layer" between the optimizer and the users - in short the users should not care what optimizations are done on the IL.
And I also don't think the optimizer should care about how the users handle the synchronization (as in, the need to run the optimizer before they access the method's body).
So this class acts as the separation layer.

What I'll do is move the hashset into here - as again the optimizer should not really care about being reentrant.

I must admit I don't like ILProvider too much, but I guess it's better than the current name.